storage/src/main/java/fk/prof/storage/buffer/StorageBackedInputStream.java

```diff
@@ -86,23 +103,35 @@ public int read(byte b[], int off, int len) throws IOException {
 
     private void fetchAndSwapBuffer() throws IOException {
         String nextFileName = fileNameStrategy.getFileName(part + 1);
+        if(buf != null) {
+            buf.close();
+        }
+        buf = null;
+
+        if(futureInputStream == null) {
+            futureInputStream = storage.fetchAsync(nextFileName);
+        }
+
         try {
-            if(buf != null) {
-                buf.close();
-            }
-            buf = storage.fetch(nextFileName);
+            buf = futureInputStream.get();
             part++;
+            futureInputStream = null;
         }
-        catch (ObjectNotFoundException e) {
-            // mark eof
-            eof = true;
-            final String msg = "File: " + nextFileName + " could not be found";
-            LOGGER.error(msg, e);
-            throw new FileNotFoundException(msg + ". Cause: " + e.getMessage());
+        catch (InterruptedException e) {
+            throw new InterruptedIOException(e.getMessage());
         }
-        catch (StorageException e) {
-            LOGGER.error("Unexpected error while fetching file: {}", nextFileName, e);
-            throw new IOException(e);
+        catch (ExecutionException e) {
+            if(e.getCause() instanceof ObjectNotFoundException) {
```

Comment:
you can simplify this as: catch (ObjectNotFoundException e1) {} catch (ExecutionException e2) {}